File: lab1-22.02.21/src/main/java/testpassword/lab1/services/UserService.java
package testpassword.lab1.services;

import lombok.val;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.mail.MailSendException;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import testpassword.lab1.models.User;
import testpassword.lab1.repos.UserRepo;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

@Service public class UserService {

    private final UserRepo repo;
    private final EmailService postman;
    private final BCryptPasswordEncoder encoder;

    @Autowired public UserService(UserRepo userRepo, EmailService postman, BCryptPasswordEncoder encoder) {
        this.repo = userRepo;
        this.postman = postman;
        this.encoder = encoder;
    }

    @Transactional public Optional<User> add(String email, String password, String name) {
        val u = new User(email, encoder.encode(password), name);
        this.save(u);
        try {
            postman.send(email, "Register", "yeah!");
        } catch (MailSendException e) { System.out.println("[ERROR] " + e.getLocalizedMessage()); }
        return this.get(email); //загружаем из базы данных, чтобы получить сущность уже с присвоенным id
    }

    @Transactional public void delete(String email) {
        repo.delete(this.repo.getByEmail(email));
        try {
            postman.send(email, "Goodbye", ":(");
        } catch (MailSendException e) { System.out.println("[ERROR] " + e.getLocalizedMessage()); }
    }

    public boolean restorePassword(String email) {
        try {
            postman.send(email, "Your password", this.get(email).get().getPassword());
            return true;
        } catch (MailSendException e) {
            System.out.println("[ERROR] " + e.getLocalizedMessage());
            return false;
        }
    }

    public List<User> getAll() { return StreamSupport.stream(repo.findAll().spliterator(), false).collect(Collectors.toList()); }

    public Optional<User> get(String email) { return Optional.ofNullable(repo.getByEmail(email)); }

    public Optional<User> get(long id) { return Optional.ofNullable(repo.getByUserId(id)); }

    public boolean exist(String email) { return repo.existsByEmail(email); }

    @Transactional public void save(User u) { this.repo.save(u); }
}

File: lab3-18.05.21/src/main/kotlin/lab3/services/UserService.kt
package lab3.services

import com.beust.klaxon.Klaxon
import lab3.models.User
import lab3.repos.UserRepo
import lab3.utils.Postman
import mu.KotlinLogging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.dao.EmptyResultDataAccessException
import org.springframework.mail.MailSendException
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*
import java.util.concurrent.TimeUnit
import javax.annotation.PostConstruct

@Service class  UserService: UserDetailsService {

    @Value("\${admin.email}") private lateinit var adminEmail: String
    @Value("\${admin.password}") private lateinit var adminPass: String
    @Value("\${admin.username}") private lateinit var adminName: String
    @Autowired private lateinit var repo: UserRepo
    @Autowired private lateinit var encoder: BCryptPasswordEncoder
    @Autowired private lateinit var postman: Postman
    @Autowired private lateinit var advertService: AdvertService
    private val log = KotlinLogging.logger {}

    @PostConstruct @Transactional
    fun initAdmin() =
        try {
            repo getByEmail adminEmail
            log.debug { "admin user already exists in db" }
        } catch (e: EmptyResultDataAccessException) {
            repo.save(User(adminEmail, encoder.encode(adminPass)).apply {
                name = adminName
                role = User.ROLE.ADMIN
            })
            log.debug { "admin user created" }
        }

    @Transactional fun add(email: String, password: String, name: String) =
        User(email, encoder.encode(password)).apply {
            repo.save(this)
            postman(email, "Register", "yeah!")
        }

    @Transactional infix fun delete(email: String) {
        repo.delete(repo.getByEmail(email))
        postman(email, "Goodbye", ":(")
    }

    @Transactional infix fun resetPassword(email: String) =
        with(loadUserByUsername(email)) {
            try {
                val tempPassword = (('a'..'z') + ('A'..'Z') + ('0'..'9')).shuffled().take(12).joinToString("")
                postman(email,
                    "Temporary password",
                    """
                    You temporary password '${tempPassword}'.
                    You should set new one on first login.
                    """.trimIndent())
                password = encoder.encode(tempPassword)
                save(this)
                true
            } catch (e: MailSendException) {
                log.error { e.message }
                false
            }
    }

    @Transactional fun modify(email: String, modified: Map<String, String>) =
        repo.save(loadUserByUsername(email).apply {
            modified["username"]?.let { name = it }
            modified["password"]?.let { password = it }
            modified["email"]?.let { this.email = it }
            modified["isDescriber"]?.let { this.isDescriber = it.toBoolean() }
        })

    // once in a week, on sunday night
    @Scheduled(cron = "* 0 0 * * 0") fun sendNewAdvertsEmails() {
        val newAdverts = advertService.getAll().filter { Date().time - it.publishDate.time > TimeUnit.HOURS.toMillis(1) }
        getAll().filter { it.isDescriber }.forEach { u ->
            postman(u.email, "New adverts available!", Klaxon().toJsonString(newAdverts.filterNot { u == it.user }))
        }
    }

    fun getAll() = repo.findAll().toList()

    infix operator fun get(id: Long) = repo getByUserId id

    override infix fun loadUserByUsername(email: String) = repo getByEmail email

    infix fun save(user: User) = repo.save(user)
}